lib/Common/Memory/PageAllocator.cpp

```diff
@@ -929,7 +929,25 @@ PageAllocatorBase<T>::FillFreePages(__in void * address, uint pageCount)
 #endif
     if (ZeroPages())
     {
-        memset(address, 0, AutoSystemInfo::PageSize * pageCount);
+        //
+        // Do memset via non-temporal store to avoid evicting existing processor cache.
+        // This helps low-end machines with limited cache size.
+        //
+#if defined(_M_IX86) || defined(_M_X64)
+        if (CONFIG_FLAG(ZeroMemoryWithNonTemporalStore))
+        {
+            size_t writeBytes = 0;
+            __m128i simdZero = _mm_setzero_si128();
+            for (__m128i * p = (__m128i *)address; writeBytes < pageCount * AutoSystemInfo::PageSize; p += 1, writeBytes += sizeof(__m128i))
+            {
+                _mm_stream_si128(p, simdZero);
+            }
+        }
+        else
+#endif
+        {
+            memset(address, 0, AutoSystemInfo::PageSize * pageCount);
+        }
```

Comment:
Somehow, api.cpp and api.h seem unrelated…  Try putting it in api.h.

From: Tom Tan [mailto:notifications@github.com]
Sent: Wednesday, July 27, 2016 6:16 PM
To: Microsoft/ChakraCore <ChakraCore@noreply.github.com>
Cc: Louis Lafreniere <louisl@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [Microsoft/ChakraCore] Zero free memory with non-temporal stores (#1311)


In lib/Common/Memory/PageAllocator.cpp<https://github.com/Microsoft/ChakraCore/pull/1311#discussion_r72551991>:

> +        //

> +#if defined(_M_IX86) || defined(_M_X64)

> +        if (CONFIG_FLAG(ZeroMemoryWithNonTemporalStore))

> +        {

> +            size_t writeBytes = 0;

> +            __m128i simdZero = _mm_setzero_si128();

> +            for (__m128i * p = (__m128i *)address; writeBytes < pageCount * AutoSystemInfo::PageSize; p += 1, writeBytes += sizeof(__m128i))

> +            {

> +                _mm_stream_si128(p, simdZero);

> +            }

> +        }

> +        else

> +#endif

> +        {

> +            memset(address, 0, AutoSystemInfo::PageSize * pageCount);

> +        }

This is good suggestion. Just try add js_memset_zero_nontemporal to api.cpp but got multiple definition of js_wmemcpy_s on Linux build. Not sure why add a reference to api.cpp in page allocator leads to such error, will revisit this later.

________________________________

In reply to: 72547225<https://github.com/Microsoft/ChakraCore/pull/1311#discussion_r72547225>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Microsoft/ChakraCore/pull/1311/files/97e55d90dcec6537e90f0a47a6272a5d66516a3d#r72551991>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AL-Bt8neBiEeZFwZ3fRiNsvitMQAKX1fks5qaALagaJpZM4JSW9g>.